File: calc/calculator-perl/calc.pl
use warnings;
use Math::Trig;
use Math::Complex;

my $result;
my $fNum;
my $sNum;

$continue = 'y';

sub add {
    my $r = $_[0] + $_[1];
    return $r;
}

sub subtract {
    my $r = $_[0] - $_[1];
    return $r;
}

sub multiply {
    my $r = $_[0] * $_[1];
    return $r;
}

sub divide {
    my $r = $_[0] / $_[1];
    return $r;
}

sub exponent {
    my $r = $_[0] ** $_[1];
    return $r;
}

sub square {
    my $r = $_[0] ** 2;
    return $r;
}

# Addition, Subtraction, Multiplication, Divisiion
# Sine (degrees), Cosine, Tangent
# Exponential, Natural Logarithm, Square Root (sqrt)
# Power of 2 (sq), Decimal Point, Positive To Negative Change (+/-)
# Clear current input (clear), cancel or reset the calculator option

while($continue eq "y") {
    print "choose operation to use:\n";
    print "1 - addition\n";
    print "2 - subtraction\n";
    print "3 - multiplication\n";
    print "4 - division\n";
    print "5 - sine\n";
    print "6 - cosine\n";
    print "7 - tangent\n";
    print "8 - Exponential\n";
    print "9 - Natural Logarithm (ln)\n";
    print "10 - Logarithm Base 10 (log)\n";
    print "11 - Power of 2 (sq)\n";
    print "12 - Square Root (sqrt)\n";

    my $operation = <STDIN>;
    chomp $operation;
    
    if($operation eq '1') {
        print "enter first number: ";
        $fNum = <STDIN>;
        chomp $fNum;
        print "enter second number: ";
        $sNum = <STDIN>;
        chomp $sNum;
        $result = add($fNum, $sNum);

        print "$fNum + $sNum = $result\n";

    } elsif ($operation eq '2') {
        print "enter first number: ";
        $fNum = <STDIN>;
        chomp $fNum;
        print "enter second number: ";
        $sNum = <STDIN>;
        chomp $sNum;
        $result = subtract($fNum, $sNum);

        print "$fNum - $sNum = $result\n";

    } elsif ($operation eq '3') {
        print "enter first number: ";
        $fNum = <STDIN>;
        chomp $fNum;
        print "enter second number: ";
        $sNum = <STDIN>;
        chomp $sNum;
        $result = multiply($fNum, $sNum);

        print "$fNum x $sNum = $result\n"

    } elsif($operation eq '4') {
        print "enter first number: ";
        $fNum = <STDIN>;
        chomp $fNum;
        print "enter second number: ";
        $sNum = <STDIN>;
        chomp $sNum;
        $result = divide($fNum, $sNum);

        print "$fNum / $sNum = $result\n"

    } elsif ($operation eq '5') {
        print "enter a number: ";
        $num = <STDIN>;
        chomp $num;
        $result = sin($num * pi / 180);

        print "sin($num) = $result\n";

    } elsif ($operation eq '6') {
        print 'enter a number: ';
        $num = <STDIN>;
        chomp $num;
        $result = cos($num * pi / 180);

        print "cos($num) = $result\n";

    } elsif ($operation eq '7') {
        print 'enter a number: ';
        $num = <STDIN>;
        chomp $num;
        $result = tan($num * pi / 180);

        print "tan($num) = $result\n";

    } elsif ($operation eq '8') {
        print 'enter first number: ';
        $fNum = <STDIN>;
        chomp $fNum;
        print 'enter second number: ';
        $sNum = <STDIN>;
        chomp $sNum;

        $result = exponent($fNum, $sNum);

        print "$fNum ^ $sNum = $result\n";

    } elsif ($operation eq '9') {
        print 'enter a number: ';
        $num = <STDIN>;
        chomp $num;
        $result = ln($num);

        print "ln($num) = $result\n";

    } elsif ($operation eq '10') {
        print 'enter a number: ';
        $num = <STDIN>;
        chomp $num;
        $result = log10($num);

        print "log10($num) = $result\n";

    } elsif ($operation eq '11') {
        print 'enter a number: ';
        $num = <STDIN>;
        chomp $num;
        $result = square($num);

        print "sq($num) = $result\n";

    } elsif ($operation eq '12') {
        print 'enter a number: ';
        $num = <STDIN>;
        chomp $num;
        $result = sqrt($num);

        print "sqrt($num) = $result\n";
    } else {
        print 'invalid operation'
    }
    print "Would you like to Continue? y/n?\n";
    $continue = <STDIN>;
    chomp $continue;
}

File: calc/calculator-html-css-js/js/calc.js

/* 

How to Write A Calculator in Javascript

Step 1: Obtain value when a button is pressed and display a number on the screen based on what you typed in.
Step 2: If you press one of the math symbol buttons, it is going to have sort of a darker frame over it to know that
you have pressed the button and you are going to do a math equation. This is going to be done in only integers as working with 
floating points would be a pain in the ass so I don't want to go down that road.

Another thing to know is that when you put a number and press one of the math symbol buttons, it is going to change back to zero
but store that value to add with the other number that is about to be used.
*/
let runningTotal = 0;
let buffer = "0";
let previousOperator = null;
const screen = document.querySelector('.screen');

function init() {
    document.querySelector(".calc-keypad").addEventListener("click", function (event) {
        buttonClick(event.target.innerText);
        // console.log('hi');
    });
}

function buttonClick(value) {
    if (isNaN(parseFloat(value))) {
        console.log(value);
        handleSymbol(value);
    } else {
        console.log(value);
        handleNumber(value);
    }
    rerender();
}

function handleNumber(value) {
    if (buffer === "0") {
        buffer = value;
    } else {
        buffer += value;
    }

}

function handleSymbol(value) {
    switch (value) {
        case 'C':
            buffer = '0';
            runningTotal = 0;
            previousOperator = null;
            break;
        case '=':
            if (previousOperator === null) {
                return;
            }
            flushOperation(parseFloat(buffer));
            previousOperator = null;
            buffer = "" + runningTotal;
            runningTotal = 0;
            break;
        case '←':
            if (buffer.length === 1) {
                buffer = "0";
            } 
            else if(buffer[0] == '-') {
                if (buffer.length == 2) {
                    buffer = "0";
                } else {
                    buffer = buffer.substring(0, buffer.length - 1);
                }
            }
            //TODO: Make it possible to back out characters when more than one value with negatives
            // else if (buffer.includes('-')) {
            //     if (buffer.length <= 2) {
            //         buffer = "0";
            //     }
            //} 
            else {
                buffer = buffer.substring(0, buffer.length - 1);
            }
            break;
        case '.':
            if (buffer.includes('.')) {
                break;
            } else {
                buffer += ".";
            }
            break;
        case '+/-':
            if (buffer[0]== '-') {
                console.log('starts with');
                buffer = buffer.substr(1);
            }
            else if (buffer == 0) {
                    break;
            }
            // else if (buffer.includes('-')) {
            //     break;
            // }
              else {
                buffer = '-' + buffer;
            }
            break;
        case 'sin':
            // console.log(buffer)
            buffer = Math.sin(buffer * Math.PI / 180);
            break;
        case 'cos':
            buffer = Math.cos(buffer * Math.PI / 180);
            break;
        case 'tan':
            buffer = Math.tan(buffer * Math.PI / 180);
            break;
        case 'sq':
            buffer = Math.pow(buffer, 2);
            break;
        case 'sqrt':
            buffer = Math.sqrt(buffer);
            break;
        case 'log':
            if (buffer == '-Infinity') {
                buffer = '0';
            }
            buffer = Math.log10(buffer);
            break;
        case 'ln':
            if (buffer == '-Infinity') {
                buffer = '0';
                break;
            }
            buffer = Math.log(buffer);
            break;
        default:
            handleMath(value);
            break;
    }
}

function handleMath(value) {
    const intBuffer = parseFloat(buffer);

    if (runningTotal === 0) {
        runningTotal = intBuffer;
    } else {
        flushOperation(intBuffer);
    }

    previousOperator = value;
    //ready for the next number to come in
    buffer = '0';
}

/*Addition, Subtraction, Multiplication, Divisiion
Sine (degrees), Cosine, Tangent
Exponential, Natural Logarithm, Square Root (sqrt)
Power of 2 (sq), Decimal Point, Positive To Negative Change (+/-)
Clear current input (clear), cancel or reset the calculator option */
function flushOperation(intBuffer) {
    if (previousOperator === "+") {
        runningTotal += intBuffer;
    } else if (previousOperator === "-") {
        runningTotal -= intBuffer;
    } else if (previousOperator === "x") {
        runningTotal *= intBuffer;
    } else if (previousOperator === "÷") {
        runningTotal /= intBuffer;
    } else if (previousOperator === "exp") {
        runningTotal = runningTotal ** intBuffer;
    }
}

function rerender(value) {
    screen.innerText = buffer;
}

init();